### PR TITLE
Initial work to support VueJS 2.0

### PR DESCRIPTION
**Pull request by petejohanson**
Here is the work so far to support VueJS 2.0. A high level overview.
- The toggle/radio controls now all work with a combination of 'v-model' and 'selected-value' properties.
- Because of the event propagation changes for Vue 2.0, Snackbar can't receive events broadcast from an arbitrary instance, so we instead have an override-able 'event-source' property. It defaults to "this.$root" so the event can be emitted on $root from any component. If you want to use some global event bus instead, you can set it there.
- There are two failing tests for badge, since, AFAICT, we can't have arbitrary extra props for directives, which was how the "hide-badge' support currently works.

Thanks!

### CODE REVIEW

File: doc/src/VmdlDoc.vue [hunk 1]
```diff
@@ -107,7 +107,7 @@ p a
 </template>
 
 <script>
-/* global docsearch*/
```
Comment (posva) [inline doc/src/VmdlDoc.vue]: can be removed

File: doc/src/VmdlDoc.vue [hunk 2]
```diff
@@ -107,7 +107,7 @@ p a
...
+/* global docsearch */
```
Comment (posva) [inline doc/src/VmdlDoc.vue]: The line can be removed

File: doc/src/main.js [hunk 3]
```diff
@@ -17,36 +17,38 @@ Vue.use(VueMdl)
...
+  routerMap.push({ path: `/${name}`, component: context(comp) })
 })
 
 let router = new VueRouter({
-})
-router.map(routerMap)
-
-router.redirect({
-  '/': '/installation'
+  routes: routerMap
```
Comment (posva) [inline doc/src/main.js]: Better call the variable `routes` and use it like `{routes}`

File: doc/src/partials/buttons.vue [hunk 4]
```diff
@@ -5,19 +5,19 @@
 .section
   title-link Buttons
   .section__content
-    p The button component allows to easily toggle classes. It supports the 
-      a(v-link.literal='/ripple-effect')
+    p The button component allows to easily toggle classes. It supports the
```
Comment (posva) [inline doc/src/partials/buttons.vue]: Careful about the extra whitespaces a the end of lines on the doc. They're actually needed. They should be replaced by `&nbsp;`

File: doc/src/partials/buttons.vue [hunk 5]
```diff
@@ -5,19 +5,19 @@
...
+      router-link(to="/ripple-effect")
         code ripple-effect
     .flex.start.wrap
       mdl-button.space Button
-      mdl-button.space(v-mdl-ripple-effect) Ripple Effect
+      mdl-button.space(mdl-ripple-effect) Ripple Effect
```
Comment (posva) [inline doc/src/partials/buttons.vue]: Why is the `v-` missing?

File: doc/src/partials/dialogs.vue [hunk 6]
```diff
@@ -7,15 +7,15 @@
   .section__content
     p The dialog component allows for verification of user actions, simple data input, and alerts to provide extra information to users. To display dialogs, use events.
 
-    div(v-transfer-dom)
-      mdl-dialog(v-ref:info-message title='Hi there')
+    div
```
Comment (posva) [inline doc/src/partials/dialogs.vue]: Argh, I guess that directive is no longer Vue 2 compatible. I'll have to find a solution to make this 2 compatible

File: doc/src/partials/dialogs.vue [hunk 7]
```diff
@@ -32,15 +32,15 @@
...
+        p= '<div>'
+        p= '  <mdl-dialog ref="infoMessage" title="Hi there">'
         p= '    <p>Hello. This is an information message. You can click outside or in the close button to close it.</p>'
         p= '  </mdl-dialog>'
         p= '</div>'
 
     p The example above use this technique to prevent the menu from overlapping.
 
-    p The dialgo component emits an&nbsp;
+    p The dialog component emits an&nbsp;
```
Comment (posva) [inline doc/src/partials/dialogs.vue]: nice catch 👍

File: doc/src/utils/hljs.js [hunk 8]
```diff
@@ -1,8 +1,12 @@
 import hljs from 'highlight.js/lib/highlight'
 
 export default {
-  update (value) {
-    this.el.innerText = value
-    hljs.highlightBlock(this.el)
+  bind (el, { value }) {
```
Comment (posva) [inline doc/src/utils/hljs.js]: no need for the bind method

File: doc/src/utils/title-link.vue [hunk 9]
```diff
@@ -22,10 +22,8 @@ a.title-link
 </style>
 
 <template lang="jade">
-a.title-link(v-link='link', :id.once='id', :class='{"title-link--big": big}')
-  h2(v-if='big')
-    slot
-  h3(v-else)
+router-link.title-link(:to='link', :id.once='id', :class="{'title-link--big': big}")
+  component(:is='header')
```
Comment (posva) [inline doc/src/utils/title-link.vue]: I can't believe this worked and I never tried it out 😆

File: src/directives/badge.js [hunk 10]
```diff
@@ -3,36 +3,28 @@ const checkNumber = function (num) {
   return num > 0
 }
 
-const dataBadgeSetter = function (hide, value) {
-  if (hide) {
-    this.el.removeAttribute('data-badge')
-  } else if (!this.isNumber || checkNumber(value)) {
-    this.el.setAttribute('data-badge', value)
+const updateBadgeValue = function (el, binding) {
+  if (!binding.modifiers.number || checkNumber(binding.value)) {
```
Comment (posva) [inline src/directives/badge.js]: Where did the hide param go?

I think we should keep this on a separate file as I plan to keep supporting vue 1 too.

File: src/directives/mdl.js [hunk 11]
```diff
@@ -1,5 +1,5 @@
 export default {
-  bind () {
-    componentHandler.upgradeElements(this.el)
+  bind (el) {
+    componentHandler.upgradeElements(el)
```
Comment (posva) [inline src/directives/mdl.js]: `componentHandler.upgradeElements(this && this.el || el)`
So we can support Vue 1

File: src/menu/menu.vue [hunk 12]
```diff
@@ -1,12 +1,12 @@
 <template lang="jade">
-ul.mdl-menu.mdl-js-menu(v-bind:for.once='for')
+ul.mdl-menu.mdl-js-menu(v-bind:for.once='target')
```
Comment (posva) [inline src/menu/menu.vue]: Why did you change this?

File: src/mixins/button.js [hunk 13]
```diff
@@ -15,33 +13,18 @@ export default {
     }
   },
   props: {
-    disabled: {
-      fill: true
-    },
+    disabled: Boolean,
```
Comment (posva) [inline src/mixins/button.js]: I'm actually surprised the propfill mixin isn't needed anymore!

File: src/mixins/prop-fill.js [hunk 14]
```diff
@@ -1,12 +1,13 @@
 // When declaring a prop with no value it should be evaluated to true
 // but '' is falsy. As a solution add the
 export default {
-  beforeCompile () {
-    if (!this._props) return
-    for (let prop of Object.keys(this._props)) {
-      let data = this._props[prop]
-      if (data.options.fill && data.raw === '') {
-        this[prop] = prop
+  created () {
```
Comment (posva) [inline src/mixins/prop-fill.js]: I prefer you keep this the way it was since is no longer used. This way I can test things easier and see if we keep it or not

File: src/mixins/toggle.js [hunk 15]
```diff
@@ -1,25 +1,35 @@
...
-        return this.checked.indexOf(this.value) >= 0
+      if (this.value instanceof Array) {
+        return this.value.indexOf(this.selectedValue) >= 0
       } else {
-        return this.checked
+        return this.value
       }
     }
+  },
+  beforeMount: function () {
```
Comment (posva) [inline src/mixins/toggle.js]: I'd prefer avoid proxying `value`. It doesn't sem necessary to me

File: src/progress.vue [hunk 16]
```diff
@@ -13,11 +11,9 @@ export default {
     buffer: {
       required: false
     },
-    indeterminate: {
-      fill: true
-    }
+    indeterminate: Boolean
```
Comment (posva) [inline src/progress.vue]: It's better to just make this kind of props take any value

File: src/select.vue [hunk 17]
```diff
@@ -35,11 +35,8 @@ export default {
     }
   },
   methods: {
-    selectValue (option) {
-      this.value = option.value
-      this.name = option.name
-      let event = new Event('change')
-      this.$el.dispatchEvent(event)
+    selectValue ({value}) {
+      this.$emit('input', value)
```
Comment (posva) [inline src/select.vue]: Having the native `change` event is still necessary. But I think I'll update this myself because it wasn't really useful without any data

File: src/select.vue [hunk 18]
```diff
@@ -82,12 +79,12 @@ export default {
...
     }
   },
-  ready () {
+  mounted () {
     componentHandler.upgradeElements(this.$el)
     this.setName()
   },
   watch: {
-    value () {
+    value: function () {
```
Comment (posva) [inline src/select.vue]: Prefer the es6 notation 😄

File: src/snackbar.vue [hunk 19]
```diff
@@ -6,11 +6,23 @@
...
+      default: function () {
+        return this.$root
+      }
+    }
+  },
+  mounted () {
     componentHandler.upgradeElement(this.$el, 'MaterialSnackbar')
-    this.$on(this.displayOn, (snackarConfig) => {
-      this.$el.MaterialSnackbar.showSnackbar(snackarConfig)
+    this.eventSource.$on(this.displayOn, (snackbarConfig) => {
```
Comment (posva) [inline src/snackbar.vue]: What is eventSource. This is also something I'll have to change. The snackback component could be better

File: src/spinner.vue [hunk 20]
```diff
@@ -1,23 +1,24 @@
 <template lang="jade">
-.mdl-spinner.mdl-js-spinner(v-bind:class='{ "mdl-spinner--single-color": singleColor, "is-active": active }')
+.mdl-spinner.mdl-js-spinner(v-bind:class="{ 'mdl-spinner--single-color': singleColor, 'is-active': active, 'is-upgraded': upgraded }")
```
Comment (posva) [inline src/spinner.vue]: I'll review this in more detail. The fact that the class is not added by MDL may be a problem.
Anyway, can you remove it from the data and class, please?

File: src/toggles/checkbox.vue [hunk 21]
```diff
@@ -1,15 +1,22 @@
 <template lang="jade">
-label.mdl-checkbox.mdl-js-checkbox(v-bind:for.once='id' v-bind:class='{ "is-disabled": disabled, "is-checked": isChecked }')
-  input.mdl-checkbox__input(v-bind:value='value' type='checkbox' v-bind:id.once='id' v-model='checked' v-bind:disabled='disabled')
+label.mdl-checkbox.mdl-js-checkbox(v-bind:for.once='id' v-bind:class="{ 'is-disabled': disabled, 'is-checked': isChecked, 'is-upgraded': upgraded }")
```
Comment (posva) [inline src/toggles/checkbox.vue]: same as before with upgraded

File: src/menu/menu.vue (hunk 12 above)
Comment (petejohanson) [inline src/menu/menu.vue]: vue-2.0.0-rc.3 gave an error if you tried to name a prop 'for' because it is a reserved JavaScript word. I just tried reverting this to test again w/ rc4, and instead it seems to just completely fail to render, without any more details as to why.

File: src/spinner.vue (hunk 20 above)
Comment (petejohanson) [inline src/spinner.vue]: The initial call to upgrade in the mounted lifecycle callback works just as expected. However, the first time you change any property that causes the binding on `class` to be recalculated, the 'is-upgraded' the MDL itself added dynamically gets removed by vue. AFAICT, it's basically the same problem reported here: http://forum.vuejs.org/topic/3963/keep-dynamically-added-class-after-re-rendering/2

I'm happy to remove it, but many of the components fail to render properly after interacting with them because 'is-upgraded' gets dropped during the vue class binding update.

File: src/snackbar.vue (hunk 19 above)
Comment (petejohanson) [inline src/snackbar.vue]: Because Vue events no longer propagate down or bubble up, The snackbar now needs to listen for events on a specific event emitter. In order to have a simple default, the property default for the source is the `$root`, which lets anyone invoke the snackbar just using `$root.$emit('mailSent', { message: 'Message Sent' })`

However, if you are using a global event bus, you might want to use that as the event emitter to listen on, in which case you could bind the `event-source` property appropriately.

I showed examples of this in the snackbar test.

**Comment by wb9688** [comment]
@petejohanson: I tried to use this, but I'm getting `[Vue warn]: You are using the runtime-only build of Vue where the template option is not available. Either pre-compile the templates into render functions, or use the compiler-included build. (found in component <MdlButton>)`, am I doing something wrong?

**Comment by posva** [comment]
@wb9688 You're requiring the wrong version of vue. Do `import vue from 'vue/dist/vue'`. That version includes the compiler

**Comment by wb9688** [comment]
@posva: When I do that, I'm getting the following errors:
```
[Vue warn]: Unknown custom element: <router-link> - did you register the component correctly? For recursive components, make sure to provide the "name" option. (found in anonymous component - use the "name" option for better debugging messages.)build.js%20line%20850%20%3E%20eval:2527:7
[Vue warn]: Unknown custom element: <mdl-button> - did you register the component correctly? For recursive components, make sure to provide the "name" option. (found in anonymous component - use the "name" option for better debugging messages.)build.js%20line%20850%20%3E%20eval:2527:7
[Vue warn]: Failed to resolve directive: mdl-ripple-effect (found in anonymous component - use the "name" option for better debugging messages.)
```
Btw is there a better way to ask for support?

**Comment by posva** [comment]
@wb9688 You have not added the plugins.
You should look at the docs. Check the readmes of vue-router and vue-mdl

**Comment by wb9688** [comment]
@posva: My code is the following:
```js
import Vue from 'vue/dist/vue'
import VueRouter from 'vue-router'
import VueMdl from 'vue-mdl'
require('material-design-lite/material.js')
require('css!material-design-lite/material.css')

Vue.use(VueRouter)
Vue.use(VueMdl)

import App from './App.vue'
import Home from './components/Home.vue'
import About from './components/About.vue'

const router = new VueRouter({
  mode: 'history',
  base: __dirname,
  routes: [
    { path: '/', component: Home },
    { path: '/about', component: About }
  ]
})

new Vue(Vue.util.extend({ router }, App)).$mount('#app')
```
`vue-router` was working fine before I changed `vue` to `vue/dist/vue`…

**Comment by posva** [comment]
@petejohanson I'm getting back to this today 😄

**Comment by petejohanson** [comment]
Thanks!

**Comment by posva** [comment]
I'm still having some bugs but most of the work is done

**Comment by juppdes** [comment]
Hello friends, there is already some provision for the release of version 2?

**Comment by posva** [comment]
@juppdes I cannot give you a date because I don't have as much time as I'd like to to spend on this project. You'll find more information at #61

**Comment by posva** [comment]
@petejohanson I integrated your changes into the v1 branch. I'm releasing a pre-release (not yet production ready) version today.
Thanks a lot for all your help with this and sorry for taking so much time. Things should go smoother now 😄

**Comment by petejohanson** [comment]
Great!

When you get to looking at MDL v2, take a look in examples/vue, I've sent them some PRs to explore how to integrate the two based on their new design. Again, nothing production worthy, but showing a working approach.

One of my biggest excitements there is the fact that they clearly are trying to support SSR approaches in the new version; I haven't yet had a chance to try implementing that in the VueJS example there.

**Comment by posva** [comment]
don't worry, I was already aware of your contributions there 😉 
As for the SSR, I didn't know, so thanks for telling me that. It's indeed something I'll have to look at once I start working on v2
